### ISSUE TITLE
[BUG]  "Completions.create() got an unexpected keyword argument 'repeat_penalty'" (related to openai-python v1?)

### ISSUE THREAD

**Issue by h0ru5** [cheshire-cat-ai/core#735]
**Describe the bug**
When using a custom openAI compatible endpoint I get an error message related to  the parameter `repeat_penalty`, which seems was dropped by the openai python package

version used is `ghcr.io/cheshire-cat-ai/core:latest`, docker image hash `08cefd8ca0c7`

**To Reproduce**
Steps to reproduce the behavior (as exampple`):
1. Setup any local OpenAI compatible endpoint (ooba, llama-cpp , candle etc.)
2.  send sny message to the cat

**Expected behavior**
it should send the prompt to the server. instead it prints an error in logs and sets the frontend into error mode

**Additional context**
seems to be related to this issue: https://github.com/openai/openai-python/issues/767#issuecomment-1805294985

```
cheshire_cat_core           | [2024-03-03 14:28:33.569] ERROR  cat.looking_glass.stray_cat.StrayCat.__call__::278
cheshire_cat_core           | "Completions.create() got an unexpected keyword argument 'repeat_penalty'"
cheshire_cat_core           | [2024-03-03 14:28:33.579] ERROR  cat.routes.websocket..websocket_endpoint::81
cheshire_cat_core           | TypeError("Completions.create() got an unexpected keyword argument 'repeat_penalty'")
cheshire_cat_core           | Traceback (most recent call last):
cheshire_cat_core           |   File "/app/cat/routes/websocket.py", line 72, in websocket_endpoint
cheshire_cat_core           |     await asyncio.gather(
cheshire_cat_core           |   File "/app/cat/routes/websocket.py", line 24, in receive_message
cheshire_cat_core           |     cat_message = await run_in_threadpool(stray, user_message)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
cheshire_cat_core           |     return await anyio.to_thread.run_sync(func, *args)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/anyio/to_thread.py", line 33, in run_sync
cheshire_cat_core           |     return await get_asynclib().run_sync_in_worker_thread(
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
cheshire_cat_core           |     return await future
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 807, in run
cheshire_cat_core           |     result = context.run(func, *args)
cheshire_cat_core           |   File "/app/cat/looking_glass/stray_cat.py", line 280, in __call__
cheshire_cat_core           |     raise e
cheshire_cat_core           |   File "/app/cat/looking_glass/stray_cat.py", line 270, in __call__
cheshire_cat_core           |     cat_message = self.agent_manager.execute_agent(self)
cheshire_cat_core           |   File "/app/cat/looking_glass/agent_manager.py", line 193, in execute_agent
cheshire_cat_core           |     out = self.execute_memory_chain(agent_input, prompt_prefix, prompt_suffix, stray)
cheshire_cat_core           |   File "/app/cat/looking_glass/agent_manager.py", line 105, in execute_memory_chain
cheshire_cat_core           |     out = memory_chain(agent_input, callbacks=[NewTokenHandler(stray)])
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain_core/_api/deprecation.py", line 145, in warning_emitting_wrapper
cheshire_cat_core           |     return wrapped(*args, **kwargs)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain/chains/base.py", line 363, in __call__
cheshire_cat_core           |     return self.invoke(
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain/chains/base.py", line 162, in invoke
cheshire_cat_core           |     raise e
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain/chains/base.py", line 156, in invoke
cheshire_cat_core           |     self._call(inputs, run_manager=run_manager)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain/chains/llm.py", line 103, in _call
cheshire_cat_core           |     response = self.generate([inputs], run_manager=run_manager)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain/chains/llm.py", line 115, in generate
cheshire_cat_core           |     return self.llm.generate_prompt(
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain_core/language_models/llms.py", line 568, in generate_prompt
cheshire_cat_core           |     return self.generate(prompt_strings, stop=stop, callbacks=callbacks, **kwargs)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain_core/language_models/llms.py", line 741, in generate
cheshire_cat_core           |     output = self._generate_helper(
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain_core/language_models/llms.py", line 605, in _generate_helper
cheshire_cat_core           |     raise e
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain_core/language_models/llms.py", line 592, in _generate_helper
cheshire_cat_core           |     self._generate(
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/langchain_openai/llms/base.py", line 340, in _generate
cheshire_cat_core           |     response = self.client.create(prompt=_prompts, **params)
cheshire_cat_core           |   File "/usr/local/lib/python3.10/site-packages/openai/_utils/_utils.py", line 271, in wrapper
cheshire_cat_core           |     return func(*args, **kwargs)
cheshire_cat_core           | TypeError: Completions.create() got an unexpected keyword argument 'repeat_penalty'
```

**Comment by valentimarco** [comment]
Hi, thanks for the report! We are aware of this issue and we are trying to resolve it, but is very difficult bc most of the service don't provide the same parameters. Maybe a list of service could resolve the issue but at the same time, create specific support using the hook [factory_allowed_llms](https://cheshire-cat-ai.github.io/docs/technical/plugins/hooks/#__tabbed_1_5) can resolve the issue and i think is the best right now!

For any update, check out our discord!

**Comment by pieroit** [comment]
@valentimarco @h0ru5 why don't we just take away `repeat_penaly` and `top_k` from the [factory](https://github.com/cheshire-cat-ai/core/blob/1e5e1a577ed5e139afab3bcd8fdf6f7da560f07c/core/cat/factory/custom_llm.py#L85)?
It's going to be less customizable but at least it will work with most APIs

If any of you two is available to make this change, please comment so I can assign the issue.
Thanks for checking this out and throw some ideas ;)

P.S.: @h0ru5 since you ar enew, remember PR branching out from `develop`

**Comment by valentimarco** [comment]
@pieroit we need to scout some info from the major providers  (TogetherAI, OpenRouters etc) and select the common params. Personally I think is better to just remove it and create integration via plugin using the hook `factory_allowed_llms`, and when one integration has more than enough downloads, then import the integration in the factory core.

What do you think?

**Comment by pieroit** [comment]
> @pieroit we need to scout some info from the major providers (TogetherAI, OpenRouters etc) and select the common params. Personally I think is better to just remove it and create integration via plugin using the hook `factory_allowed_llms`, and when one integration has more than enough downloads, then import the integration in the factory core.
> 
> Or, just put 4 params: base_url, temperature, model_name, api_key. But we lose the "control" of the model (right now is very important and most providers give different level of control)
> 
> What do you think?

Agree on having the specialized clients as plugins, but why do you see a problem with a minimal OpenaAICompatible client? Without params only the address?

**Comment by valentimarco** [comment]
> > @pieroit we need to scout some info from the major providers (TogetherAI, OpenRouters etc) and select the common params. Personally I think is better to just remove it and create integration via plugin using the hook `factory_allowed_llms`, and when one integration has more than enough downloads, then import the integration in the factory core.
> > Or, just put 4 params: base_url, temperature, model_name, api_key. But we lose the "control" of the model (right now is very important and most providers give different level of control)
> > What do you think?
> 
> Agree on having the specialized clients as plugins, but why do you see a problem with a minimal OpenaAICompatible client? With params only the strictly necessary params?

Well, most open models need some "parmas fine-tuning" to have the best possible...
We can just refactor to give a minimal client and if ppl needs something more specific, than use the plugin / create one

**Comment by pieroit** [comment]
Closing this as we will fix in #759
